target/linux/ramips/dts/TL-WR840NV4.dts

```diff
@@ -0,0 +1,141 @@
+/dts-v1/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+#include "mt7628an.dtsi"
+
+/ {
+	compatible = "tplink,tl-wr840n-v4", "mediatek,mt7628an-soc";
+	model = "TP-LINK TL-WR840N v4";
+
+	chosen {
+		bootargs = "console=ttyS0,115200";
+	};
+
+	memory@0 {
+		device_type = "memory";
+		reg = <0x0 0x4000000>;
+	};
+
+	gpio-keys-polled {
+		compatible = "gpio-keys-polled";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		poll-interval = <20>;
+
+		reset {
+			label = "reset";
+			gpios = <&gpio1 6 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+	};
+
+	gpio-leds {
+		compatible = "gpio-leds";
+
+		power	 {
+			label = "tl-wr840n-v4:green:power";
+			gpios = <&gpio1 4 GPIO_ACTIVE_LOW>;
+		};
+
+		wireless {
+			label = "tl-wr840n-v4:green:wireless";
+			gpios = <&gpio1 12 GPIO_ACTIVE_LOW>;
+		};
+
+		ethernet {
+			label = "tl-wr840n-v4:green:lan";
+			gpios = <&gpio1 9 GPIO_ACTIVE_LOW>;
+		};
+
+		internet {
+			label = "tl-wr840n-v4:green:wan";
+			gpios = <&gpio1 11 GPIO_ACTIVE_LOW>;
+		};
+
+		wps {
+			label = "tl-wr840n-v4:green:wps";
+			gpios = <&gpio1 5 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&pinctrl {
+	state_default: pinctrl0 {
+		gpio {
+			ralink,group = "i2c", "i2s", "uart1", "uart2", "p0led_an", "p2led_an" ;
+			ralink,function = "gpio";
+		};
+	};
+};
+
+&spi0 {
+	status = "okay";
+
+	m25p80@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
+		m25p,chunked-io = <32>;
+
+		partition@0 {
+			label = "boot";
+			reg = <0x0 0x20000>;
+			read-only;
+		};
+
+		partition@20000 {
+			label = "firmware";
+			reg = <0x20000 0x7a0000>;
+		};
+
+		partition@7c0000 {
+			label = "config";
+			reg = <0x7c0000 0x10000>;
+			read-only;
+		};
+
+		factory: partition@07d0000 {
+			label = "romfile";
+			reg = <0x7d0000 0x10000>;
+			read-only;
+		};
+
+		partition@07e0000 {
+			label = "rom";
+			reg = <0x7e0000 0x10000>;
+			read-only;
+		};
+
+		partition@07f0000 {
+			label = "radio";
+			reg = <0x7f0000 0x10000>;
+			read-only;
+		};
+	};
+};
+
+&usbphy {
+	status = "disabled";
+};
+
+&ehci {
+	status = "disabled";
+};
+
+&ohci {
+	status = "disabled";
+};
+
+&wmac {
+	status = "okay";
+	mtd-mac-address = <&factory 0xf100>;
```

Comment:
I'm trying to investigate poor wireless performance, which was described by @sanitariu 
The thing I noticed here is that the EEPROM address was not defined, so it uses default value from mt7628an.dtsi, which is wrong for this hardware:
`mediatek,mtd-eeprom = <&factory 0x0000>;`
Real EEPROM is located in radio partition, so we must declare this:
```
&spi0 {
...
	m25p80@0 {
...
		radio: partition@07f0000 {
...
&wmac {
...
	mediatek,mtd-eeprom = <&radio 0x0000>;
};
```

But performance is still bad though.